tempodb/tempodb.go

```diff
@@ -309,77 +319,102 @@ func (rw *readerWriter) pollBlocklist() {
+			continue
 		}
 
-		interfaceSlice := make([]interface{}, 0, len(blockIDs))
-		for _, id := range blockIDs {
-			interfaceSlice = append(interfaceSlice, id)
-		}
+		metricBlocklistLength.WithLabelValues(tenantID).Set(float64(len(newBlockList)))
 
-		listMutex := sync.Mutex{}
-		blocklist := make([]*backend.BlockMeta, 0, len(blockIDs))
-		compactedBlocklist := make([]*backend.CompactedBlockMeta, 0, len(blockIDs))
-		_, err = rw.pool.RunJobs(ctx, interfaceSlice, func(ctx context.Context, payload interface{}) ([]byte, error) {
-			blockID := payload.(uuid.UUID)
-
-			var compactedBlockMeta *backend.CompactedBlockMeta
-			blockMeta, err := rw.r.BlockMeta(ctx, blockID, tenantID)
-			// if the normal meta doesn't exist maybe it's compacted.
-			if err == backend.ErrMetaDoesNotExist {
-				blockMeta = nil
-				compactedBlockMeta, err = rw.c.CompactedBlockMeta(blockID, tenantID)
-			}
+		rw.blockListsMtx.Lock()
+		rw.blockLists[tenantID] = newBlockList
+		rw.compactedBlockLists[tenantID] = newCompactedBlockList
+		rw.blockListsMtx.Unlock()
+	}
+}
 
-			// blocks in intermediate states may not have a compacted or normal block meta.
-			//   this is not necessarily an error, just bail out
-			if err == backend.ErrMetaDoesNotExist {
-				return nil, nil
-			}
+func (rw *readerWriter) pollTenant(ctx context.Context, tenantID string) ([]*backend.BlockMeta, []*backend.CompactedBlockMeta, error) {
+	blockIDs, err := rw.r.Blocks(ctx, tenantID)
+	if err != nil {
+		metricBlocklistErrors.WithLabelValues(tenantID).Inc()
+		level.Error(rw.logger).Log("msg", "error polling blocklist", "tenantID", tenantID, "err", err)
+		return nil, nil, err
+	}
 
-			if err != nil {
-				metricBlocklistErrors.WithLabelValues(tenantID).Inc()
-				level.Error(rw.logger).Log("msg", "failed to retrieve block meta", "tenantID", tenantID, "blockID", blockID, "err", err)
-				return nil, nil
+	bg := tempoUtil.NewBoundedWaitGroup(rw.cfg.BlocklistPollConcurrency)
+	chErr := make(chan error, len(blockIDs))
+	chMeta := make(chan *backend.BlockMeta, len(blockIDs))
+	chCompactedMeta := make(chan *backend.CompactedBlockMeta, len(blockIDs))
+
+	for _, blockID := range blockIDs {
+		bg.Add(1)
+		go func(b uuid.UUID) {
+			defer bg.Done()
+			m, cm, e := rw.pollBlock(ctx, tenantID, b)
+			if e != nil {
+				chErr <- e
+			} else if m != nil {
+				chMeta <- m
+			} else if cm != nil {
+				chCompactedMeta <- cm
 			}
+		}(blockID)
+	}
 
-			// todo:  make this not terrible. this mutex is dumb we should be returning results with a channel. shoehorning this into the worker pool is silly.
-			//        make the worker pool more generic? and reusable in this case
-			listMutex.Lock()
-			if blockMeta != nil {
-				blocklist = append(blocklist, blockMeta)
+	bg.Wait()
+	close(chErr)
+	close(chMeta)
+	close(chCompactedMeta)
 
-			} else if compactedBlockMeta != nil {
-				compactedBlocklist = append(compactedBlocklist, compactedBlockMeta)
-			}
-			listMutex.Unlock()
+	for err := range chErr {
```

Comment:
Since this was updated to continue despite an error polling, there is no need to capture errors anymore and this was removed.  I don't think we even need an atomic single error here, the errors are caught and logged/metriced in the child functions and not propagated up.